lib/options.go

```diff
@@ -187,6 +189,117 @@ func ParseCIDR(s string) (*IPNet, error) {
 	return &parsedIPNet, nil
 }
 
+// HostnameTrie is a tree-structured list of hostname matches with support
+// for wildcards exclusively at the start of the pattern. Items may only
+// be inserted and searched. Internationalized hostnames are valid.
+type HostnameTrie struct {
+	children []*HostnameTrie
+	r        rune
+	terminal bool // end of a valid match
+}
+
+// Regex description of hostname pattern to enforce blocks by. Global var
+// to avoid compilation penalty at runtime.
+// Matches against strings composed entirely of letters, numbers, or '.'s
+// with an optional wildcard at the start.
+//nolint:gochecknoglobals
+var legalHostnamePattern *regexp.Regexp = regexp.MustCompile(`^\*?(\pL|[0-9\.])*`)
+
+func legalHostname(s string) error {
+	if len(legalHostnamePattern.FindString(s)) != len(s) {
+		return errors.Errorf("invalid hostname pattern %s", s)
+	}
+	return nil
+}
+
+// UnmarshalJSON forms a HostnameTrie from the provided hostname pattern
+// list.
+func (t *HostnameTrie) UnmarshalJSON(data []byte) error {
+	m := make([]string, 0)
+	if err := json.Unmarshal(data, &m); err != nil {
+		return err
+	}
+	for _, h := range m {
+		if insertErr := t.Insert(h); insertErr != nil {
+			return insertErr
+		}
+	}
+	return nil
+}
+
+// UnmarshalText forms a HostnameTrie from a comma-delimited list
+// of hostname patterns.
+func (t *HostnameTrie) UnmarshalText(b []byte) error {
+	for _, s := range strings.Split(string(b), ",") {
+		if err := t.Insert(s); err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+// Insert a hostname pattern into the given HostnameTrie. Returns an error
+// if hostname pattern is illegal.
+func (t *HostnameTrie) Insert(s string) error {
+	if len(s) == 0 {
+		return nil
+	}
+
+	if err := legalHostname(s); err != nil {
+		return err
+	}
+
+	rStr := []rune(s) // need to iterate by runes for intl' names
+	last := len(rStr) - 1
+	for _, c := range t.children {
+		if c.r == rStr[last] {
```

Comment:
the biggest benefit would from the above suggestion will probably be that you could just do something like
```go
if c, ok := t.children[rStr[last]]; ok {
   return c.Insert(string(rStr[:last]))
}
```